pyuvdata/uvcal.py

```diff
@@ -0,0 +1,289 @@
+                                                 expected_type=np.bool)
+
+        desc = ('Origin (on github for e.g) of calibration software. ')
+        self._git_origin = uvp.UVParameter('git_origin', form='str',
+                                           expected_type=str,
+                                           description=desc,
+                                           required=False)
+
+        desc = ('Commit hash of calibration software (from git_origin) used'
+                'to generate solutions.')
+        self._git_hash = uvp.UVParameter('git_hash', form='str',
+                                         expected_type=str,
+                                         description=desc,
+                                         required=False)
+
+        decs = ('Name of calibration pipeline, e.g. Omnical. String.')
+        self._pipeline = uvp.UVParameter('pipeline', form='str',
+                                         expected_type=str,
+                                         description=desc,
+                                         required=False)
+
+        desc = ('Name of observer who calculated solutions in this file.')
+        self._observer = uvp.UVParameter('observer',
+                                               description=desc,
+                                               form='str',
+                                               expected_type=str,
+                                               required=False)
+
+
+        super(UVCal, self).__init__()
+
+    def check(self, run_check_acceptability=True):
+        """
+        Add some extra checks on top of checks on UVBase class.
+
+        Check that all required parameters are set reasonably.
+
+        Check that required parameters exist and have appropriate shapes.
+        Optionally check if the values are acceptable.
+
+        Args:
+            run_check_acceptability: Option to check if values in required parameters
+                are acceptable. Default is True.
+        """
+        # first run the basic check from UVBase
+        super(UVCal, self).check(run_check_acceptability=run_check_acceptability)
+
+        # then check some other things
+        nants_data_calc = int(len(np.unique(self.antenna_numbers)))
+        if self.Nants_data != nants_data_calc:
+            raise ValueError('Nants_data must be equal to the number of unique '
+                             'values antenna_numbers.')
+
+        if self.Nants_data > self.Nants_telescope:
+            raise ValueError('Nants_data must be less than or equal to Nants_telescope')
+        return True
+
+    def set_gain(self):
+        """Set cal_type to 'gain' and adjust required parameters."""
+        self.cal_type = 'gain'
+        self._gain_array.required = True
+        self._delay_array.required = False
+        self._quality_array.form = self._gain_array.form
+        self._flag_array.form = self._gain_array.form
```

Comment:
I don't think you need this line in any of the set_* methods because it's always the same and none of them change it. You need the quality one because it's different for the different cal types